### PR TITLE
Lcov report improvements part 2

### PR DESCRIPTION
**Pull request by zackw**
- break up lcov_file to not be so gigantic
- improved branch destination reporting
- function coverage!
- addresses #1850 and #1852

Because of the refactor in the first commit, functional changes will be easier to see if you go commit by commit.

### CODE REVIEW

**Comment by zackw** [comment]
I think this is ready to be reviewed now.  Let me know if you want me to clean up the commit history or anything.

**Comment by nedbat** [comment]
BTW, I'm about to drop Python 3.8, so you can simplify the tests by just skipping them on pypy3.8.

**Comment by zackw** [comment]
Testing whether the existing `@xfail_pypy38` decorator does the trick now.

**Comment by nedbat** [comment]
Looks good, thanks so much! I'll merge it.

**Comment by nedbat** [comment]
This is now released as part of [coverage 7.6.2](https://pypi.org/project/coverage/7.6.2).

File: coverage/lcovreport.py [hunk 1]
```diff
@@ -32,6 +32,129 @@ def line_hash(line: str) -> str:
...
+        # and not-executed arcs to BRDA codes.  It seems to produce reasonable
+        # results when fed through genhtml.
+        _, taken = branch_stats[line]
+
+        if taken == 0:
+            # When _none_ of the out arcs from 'line' were executed,
+            # this probably means 'line' was never executed at all.
+            # Cross-check with the line stats.
+            assert len(executed_arcs[line]) == 0
+            assert line in analysis.missing
```
Comment (bgianfo) [inline coverage/lcovreport.py:123]: Our coverage workflow is now failing due to this line, when running with version `7.6.3`.

Comment (nedbat) [reply coverage/lcovreport.py:123]: Can you provide us with a way to reproduce it?  Failing that, can you add this line before the assert, and show us the output?
```py
print(f"{line=}, {analysis.missing=}")
```
(this should have been on the assert line, my bad)

Comment (rouge8) [reply coverage/lcovreport.py:123]: I'm seeing this same error too, but only in CI so far. I haven't been able to reproduce it locally so I can't put the `print()` line in either...

**Pull request merged**